File: python/src/chauff_cmn/logging/_correlation.py
from contextvars import ContextVar, Token
from typing import Mapping, Optional
from uuid import uuid4

REQUEST_ID_HEADER = "X-Request-Id"

# Portée par requête: posée par le middleware au début de la requête, lue par
# le patcher loguru (voir __init__.py) pour injecter correlation_id dans
# CHAQUE ligne de log émise pendant le traitement, pas seulement celle du
# middleware lui-même.
correlation_id_var: ContextVar[Optional[str]] = ContextVar("chauff_cmn_correlation_id", default=None)


def resolve_correlation_id(headers: Mapping[str, str]) -> str:
    """Réutilise l'id de corrélation entrant s'il existe, sinon en génère un.

    Fonctionne avec un CIMultiDict aiohttp (clés insensibles à la casse) et un
    dict brut d'en-têtes ASGI (clés déjà en minuscules par spec), d'où la
    comparaison manuelle plutôt qu'un simple `.get(REQUEST_ID_HEADER)`.
    """
    for key, value in headers.items():
        if key.lower() == REQUEST_ID_HEADER.lower() and value:
            return value
    return str(uuid4())


def bind_correlation_id(value: str) -> Token:
    return correlation_id_var.set(value)


def reset_correlation_id(token: Token) -> None:
    correlation_id_var.reset(token)

File: typescript/src/logging/_correlation.ts
import { AsyncLocalStorage } from "node:async_hooks";

export const REQUEST_ID_HEADER = "X-Request-Id";

interface CorrelationStore {
  correlationId: string;
}

// Portée par requête, façon contextvar Python : posée par le call site qui
// englobe une requête (ex. withRequestLogging), lue par write() dans
// index.ts pour injecter correlation_id dans CHAQUE ligne de log émise
// pendant le traitement, pas seulement une ligne dédiée au middleware.
const storage = new AsyncLocalStorage<CorrelationStore>();

type HeaderSource = Headers | Record<string, string | string[] | undefined>;

function readHeader(headers: HeaderSource, name: string): string | undefined {
  if (typeof (headers as Headers).get === "function") {
    return (headers as Headers).get(name) ?? undefined;
  }
  const record = headers as Record<string, string | string[] | undefined>;
  for (const key of Object.keys(record)) {
    if (key.toLowerCase() === name.toLowerCase()) {
      const value = record[key];
      return Array.isArray(value) ? value[0] : value;
    }
  }
  return undefined;
}

// Réutilise l'id de corrélation entrant s'il existe, sinon en génère un —
// équivalent de resolve_correlation_id côté Python.
export function resolveCorrelationId(headers: HeaderSource): string {
  const existing = readHeader(headers, REQUEST_ID_HEADER);
  return existing && existing.length > 0 ? existing : crypto.randomUUID();
}

export function runWithCorrelationId<T>(correlationId: string, fn: () => T): T {
  return storage.run({ correlationId }, fn);
}

export function getCorrelationId(): string | null {
  return storage.getStore()?.correlationId ?? null;
}
